### PR TITLE
[WIP] Realtime API

### PR DESCRIPTION
**Pull request by ian-r-rose**
Selective merge of implementation-agnostic parts for a realtime collaboration service. Currently not hooked up to anything, but I have an implementation using Google drive.
In particular, the interface is described in `realtime/realtime.ts`.

I am still getting the service/plugin architecture ironed out, so bear with me.

The specific implentation using Google Drive+Realtime can be found:
https://github.com/ian-r-rose/jupyterlab_google_drive

### CODE REVIEW

**Comment by blink1073** [comment]
This is great!  What if we updated `IObservableVector` and `IObservableString` themselves for real-time handling?  That way, the models using them would not have to worry about lifecycle issues like signal handling and disposal as real-time is enabled or disabled.

**Comment by blink1073** [comment]
So `createString` would become `handleString(value: IObservableString): void`.

**Comment by ian-r-rose** [comment]
That would probably be simpler overall. My initial implementations of this were closer to that, in fact, but I moved over to the current pattern. Basically, the internal implementations for the realtime and non-realtime versions of these objects are different enough that I could not see a simple way to swap them out while maintaining the lifecycle issues.

Ultimately, it seemed simpler to have the `IRealtimeModel` objects to reconnect the relevant signals to a new object. Or do I misunderstand your suggestion?

**Comment by blink1073** [comment]
I meant this block:

```typescript
registerCollaborative( realtimeHandler : IRealtimeHandler ) : Promise<void> {
    return new Promise<void>((resolve,reject)=>{
      this._realtime = realtimeHandler;

      //create a new realtime string
      this._realtime.createString(this._text.text).then((str: IObservableString)=>{
        let oldStr = this._text;
        this._text = str;
        //connect the realtime string to the correct signals
        this._text.changed.connect(()=>{
          this.contentChanged.emit(void 0);
          this.dirty = true;
        });
        //get rid of the old string.
        oldStr.dispose();
        resolve();
      }).catch(()=>{
        console.log("Unable to register document as collaborative");
      });
    });
}
```

Would become:

```typescript
registerCollaborative( realtimeHandler : IRealtimeHandler ) : Promise<void> {
    return realtimeHandler.handleString(this._text).then(value => {
        if (!value) {
            console.log("Unable to register document as collaborative");
        }
    });
}
```

**Comment by ian-r-rose** [comment]
Yes, but I would still need to create a new `IObservableString` and attach it to the `this` object, no? That is to say, the `handleString()` pseudocode would be something like

```typescript
handleString( text: IObservableString ) : Promise<boolean> {
  text = new RealtimeString(text);
  cloneSignals(text);
  resolve(true);
}
```

**Comment by blink1073** [comment]
No, I'm saying that the `IObservableString` *itself* becomes real-time enabled.  It would clear its internal `_text` and synchronize itself with the realtime string.

**Comment by blink1073** [comment]
So that the model using it can continue to do so without worrying about that transition.

**Comment by blink1073** [comment]
We'd add a method to `IObservableString` that accepts a `RealtimeString` to sync up with.

**Comment by ian-r-rose** [comment]
I see. My issue with that is that the swapping becomes extremely implementation-dependent. That is, a realtime string using the Google API might be completely different from another (like if we have a realtime string syncing through the Jupyter server).

**Comment by blink1073** [comment]
I just thought of another reason we can't swap out the observable on the model: we don't know who else has connected to the existing observable.  We have to either keep both and have the `realtimeString` sync itself with the `IObservableString` or find a way to optimize it so we don't store the same value in both places.

**Comment by blink1073** [comment]
Maybe the method is something like `link(parent: IObservableString)`.  This would internally clear its text and sync and slave itself to the parent.

**Comment by ian-r-rose** [comment]
A very good point. I would point out that having the reference implementation of the `IObservableString` and `IObservableVector` interfaces be able to themselves sync to some sort of `Realtime*` object is that then these reference implementations would become significantly more complex. Not to mention another layer of indirection with defining `Realtime*` (rather than just implementing another `IObservable*`)

**Comment by blink1073** [comment]
That method could return a disposable that clears an internal `_linked` flag, removes the signals, and sets its internal `_text` to the value of the parent.

**Comment by blink1073** [comment]
This interface presumes no realtime.

**Comment by ian-r-rose** [comment]
You are right, that presumes no realtime. I could see that working...

**Comment by blink1073** [comment]
Nice!  I'm exciting about getting this and the changes from #1419.

**Comment by ian-r-rose** [comment]
Now, consider the more complex `IObservableVector` case. Would every member function of that object have some sort of test 
```typescript
if (linked) {
  return this._parent.*()
} else {
  return this.*()
}
```
?
Not necessarily the most elegant, but maybe fine.

**Comment by blink1073** [comment]
I think it would have to.

**Comment by ian-r-rose** [comment]
Alternatively, the `IObservable*` classes could have a boolean readonly `linkable`.

**Comment by blink1073** [comment]
The linkable property sounds good.

**Comment by jasongrout** [comment]
Commenting just on the name-how about isLinkable? That's more in line with phosphor convention

**Comment by blink1073** [comment]
Heads up, there are some merge conflicts due to some refactoring this past week.

**Comment by ian-r-rose** [comment]
Things are moving fast, I just did a big rebase yesterday!

**Comment by blink1073** [comment]
Just one more to go: https://github.com/jupyterlab/jupyterlab/issues/1364

**Comment by ian-r-rose** [comment]
By the way, I am planning to make some changes to the overall Realtime services API, but I could use some eyes on the linking semantics for the ObservableVector/ObservableString.

**Comment by blink1073** [comment]
@ian-r-rose, I've been following along with your commits.  Everything looks good to me thus far.

File: src/realtime/realtime.ts [hunk 1]
```diff
@@ -157,6 +157,20 @@ interface IRealtimeHandler extends IDisposable {
   readonly collaborators: IObservableMap<ICollaborator>;
 
   /**
+   * Create a vector for the realtime model.
```
Comment (blink1073) [inline src/realtime/realtime.ts]: vector -> map?

**Comment by blink1073** [comment]
Don't forget to add the new test file to the test index.

**Comment by ian-r-rose** [comment]
Whoops, thanks.

**Comment by ellisonbg** [comment]
...rebase...

**Comment by blink1073** [comment]
Hi @ian-r-rose, I've written up some [notes](https://gist.github.com/blink1073/fe1e8505c2caf084593cffa611bfb90d) with some ideas about how we can rectify realtime and local storage.  Are you available tomorrow to discuss?

**Comment by ian-r-rose** [comment]
Absolutely. I should be pretty flexible, so feel free to suggest a time.

**Comment by blink1073** [comment]
1000 Pacific?

**Comment by ian-r-rose** [comment]
Works for me. Talk to you then.

**Comment by blink1073** [comment]
Notes from that discussion [here](https://gist.github.com/blink1073/fe1e8505c2caf084593cffa611bfb90d#notes-from-conversation-with-chris-and-ian-on-22-feb)

**Comment by ellisonbg** [comment]
@ian-r-rose can we open an issue to use to track the design discussion about RTC? That is more accessible than a gist. and it is easy for others to contribute. Thanks!

**Comment by ian-r-rose** [comment]
Closed in favor of #2021.

**Pull request closed without merging**